File: offer/JZ8 二叉树的下一个结点/tree.java
/*
public class TreeLinkNode {
    int val;
    TreeLinkNode left = null;
    TreeLinkNode right = null;
    TreeLinkNode next = null;

    TreeLinkNode(int val) {
        this.val = val;
    }
}
*/
public class Solution {
    public TreeLinkNode GetNext(TreeLinkNode root) {
        if(root == null)
            return null;
        if(root.right != null){
            root = root.right;
            while(root.left != null){
                root = root.left;
            }
            return root;
        }
        while(root.next !=  null){
            TreeLinkNode parent = root.next;
            if(parent.left == root){
                return parent;
            }
            root = root.next;
        }
        return null;
    }
}

File: 0114. Flatten Binary Tree to Linked List/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        adjust(root);
    }
    
    TreeNode* adjust(TreeNode* root){
        if(root == nullptr)
            return nullptr;
        TreeNode *left = adjust(root->left), *right = adjust(root->right);
        if(root->left != nullptr){
            TreeNode* oldRight = root->right;
            root->right = root->left;
            root->left = nullptr;
            left->right = oldRight;
        }
        if(right != nullptr){
            return right;
        }
        if(left != nullptr){
            return left;
        }
        return root;
    }
};